flash/flash.go

```diff
@@ -0,0 +1,370 @@
+			}
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// disable sector protection if the chip has it
+	if dev.attrs.HasSectorProtection {
+		if err := dev.WriteEnable(); err != nil {
+			return err
+		}
+		if err := dev.trans.writeCommand(cmdWriteStatus, []byte{0x00}); err != nil {
+			return err
+		}
+	}
+
+	// write disable
+	if err := dev.trans.runCommand(cmdWriteDisable); err != nil {
+		return err
+	}
+	return dev.WaitUntilReady()
+}
+
+// Attrs returns the attributes of the device determined from the most recent
+// call to Configure(). If no call to Configure() has been made, this will be
+// the zero value of the Attrs struct.
+func (dev *Device) Attrs() Attrs {
+	return dev.attrs
+}
+
+// ReadJEDEC reads the JEDEC ID from the device; this ID can then be used to
+// ascertain the attributes of the chip from a list of known devices.
+func (dev *Device) ReadJEDEC() (JedecID, error) {
+	jedecID := make([]byte, 3)
+	if err := dev.trans.readCommand(cmdReadJedecID, jedecID); err != nil {
+		return JedecID{}, err
+	}
+	return JedecID{jedecID[0], jedecID[1], jedecID[2]}, nil
+}
+
+// ReadSerialNumber reads the serial numbers from the connected device.
+func (dev *Device) ReadSerialNumber() (SerialNumber, error) {
+	sn := make([]byte, 12)
+	if err := dev.trans.readCommand(0x4B, sn); err != nil {
+		return 0, err
+	}
+	return SerialNumber(uint64(sn[11]) | uint64(sn[10])<<0x8 |
+		uint64(sn[9])<<0x10 | uint64(sn[8])<<0x18 | uint64(sn[7])<<0x20 |
+		uint64(sn[6])<<0x28 | uint64(sn[5])<<0x30 | uint64(sn[4])<<0x38), nil
+}
+
+// ReadBuffer fills the provided buffer with memory read from the device
+// starting at the provided address.
+func (dev *Device) ReadBuffer(addr uint32, buf []byte) error {
+	if err := dev.WaitUntilReady(); err != nil {
+		return err
+	}
+	return dev.trans.readMemory(addr, buf)
+}
+
+// WriteBuffer writes data to the device, one page at a time, starting at the
+// provided address. This method assumes that the destination is already erased.
+func (dev *Device) WriteBuffer(addr uint32, buf []byte) (n int, err error) {
+	remain := uint32(len(buf))
+	idx := uint32(0)
+	for remain > 0 {
+		if err = dev.WaitUntilReady(); err != nil {
+			return
+		}
+		if err = dev.WriteEnable(); err != nil {
+			return
+		}
+		leftOnPage := PageSize - (addr & (PageSize - 1))
+		toWrite := remain
+		if leftOnPage < remain {
+			toWrite = leftOnPage
+		}
+		if err = dev.trans.writeMemory(addr, buf[idx:idx+toWrite]); err != nil {
+			return
+		}
+		idx += toWrite
+		addr += toWrite
+		remain -= toWrite
+	}
+	return len(buf) - int(remain), nil
+}
+
+func (dev *Device) WriteEnable() error {
+	return dev.trans.runCommand(cmdWriteEnable)
+}
+
+// EraseBlock erases a block of memory at the specified address
+func (dev *Device) EraseBlock(addr uint32) error {
+	if err := dev.WaitUntilReady(); err != nil {
+		return err
+	}
+	if err := dev.WriteEnable(); err != nil {
+		return err
+	}
+	return dev.trans.eraseCommand(cmdEraseBlock, addr*BlockSize)
```

Comment:
It looks like `addr` isn't a real address but instead is a block number?
Same goes for `EraseSector`.

Reply:
You're right, this is likely a bug, and certainly the variable name is misleading.  I will double check the functionality and fix up accordingly.